webservices/resources/legal.py

```diff
@@ -22,6 +24,45 @@ def get(self, ao_no, **kwargs):
         return results
 
 
+phrase_regex = re.compile('"(?P<phrase>[^"]*)"')
+def parse_query_string(query):
+    """Parse phrases from a query string for exact matches e.g. "independent agency"."""
+
+    def _parse_query_string(query):
+        """Recursively pull out terms and phrases from query. Each pass pulls
+        out terms leading up to the phrase as well as the phrase itself. Then
+        it processes the remaining string."""
+
+        if not query:
+            return ([], [])
+
+        match = phrase_regex.search(query)
```

Comment:
Short answer is that it seemed like a complicated regex/looping problem thinking in recursion terms helped me reason through it. It's really just a loop that is chunking over the query string using regex to chunk it up.

`findall` doesn't return a list of matches, it returns strings or tuples, so you loose the nice api from regex matches. In my head, going with `for` and `findall`, you would have a long loop (in terms of lines of code within the block, my current solution is pushing what I would want in a `for` loop to give you an idea) to split the terms from the phrases. To do that you would have a variable with the remaining string, or be using start/end pointers to keep track of your work. OR you could have a slightly more complex regex, I was thinking `((?P<terms>[^\s"]+)\s+)?"(?P<phrase>[^"]*)"` and dealing with the groups as tuples wouldn't be so fun. The regex is hard here, because the query string is pretty free-form so I was hoping to keep it simple. When looping, if you've got more than one or two variables mutating, my head just goes to a function at that point with immutable variables to reduce the complexity. This is entirely subjective, and I didn't actually write the code in the `for`/`while` loop style, so I don't really know how it would've looked.

BUT, in general, I'm all for learning new APIs and getting simpler, more readable, efficient code in there, so feel free to post PR coments with code snippets or pseudo code if you have something in mind that would work better. I certainly do that to other folks as well, at least to get their opinion on my snippets :)

Thinking about it again, something like this might be better, but I would definitely want to add more test cases. IDK, what do you think?
```python
    def _parse_query_string(query):
        terms = []
        phrases = []

        matches = re.findall('((?P<term>[^\s"]+)*\s*)("(?P<phrase>[^"]*)")?', query)
        for match in matches:
            _, term, _, phrase = match
            if term:
                terms.append(term)
            if phrase:
                phrases.append(phrase)

        return (terms, phrases)
```